src/v/cluster/topics_frontend.cc

```diff
@@ -949,4 +979,163 @@ topics_frontend::do_cancel_moving_partition_replicas(
     co_return results;
 }
 
+ss::future<std::error_code> topics_frontend::change_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor new_replication_factor,
+  model::timeout_clock::time_point timeout) {
+    auto tp_metadata = _topics.local().get_topic_metadata_ref(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+    if (!tp_metadata.value().get().is_topic_replicable()) {
+        co_return errc::topic_operation_error;
+    }
+
+    auto current_replication_factor
+      = tp_metadata.value().get().get_replication_factor();
+
+    if (current_replication_factor == new_replication_factor) {
+        co_return errc::success;
+    }
+
+    if (new_replication_factor < current_replication_factor) {
+        co_return co_await decrease_replication_factor(
+          topic, new_replication_factor, timeout);
+    }
+    co_return co_await increase_replication_factor(
+      topic, new_replication_factor, timeout);
+}
+
+ss::future<std::error_code> topics_frontend::increase_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor new_replication_factor,
+  model::timeout_clock::time_point timeout) {
+    std::vector<move_topic_replicas_data> new_assignments;
+
+    auto tp_metadata = _topics.local().get_topic_metadata(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+
+    auto partition_count = tp_metadata->get_configuration().partition_count;
+
+    // units shold exist during replicate_and_wait call
+    std::vector<allocation_units> units;
```

Comment:
i think we need to keep `ss::foreign_ptr` here as the units will always come from core 0 and frontend lives on every core